Added pagination support for /user/teams github api with link header .

toshi-miura opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

Added handling of last page pointed out below to [#102](https://github.com/pusher/oauth2_proxy/pull/102)
>Could we maybe use the Link values to make sure we only call up until the last page?

Test is done by rewriting the code with a smaller "per_page" value.

JoelSpeed commented on providers/github.go:
```diff
@@ -139,36 +140,95 @@ func (p *GitHubProvider) hasOrgAndTeam(accessToken string) (bool, error) {
+			// 1. When paging is required (Example: When the data size is 100 and the page size is 99 or less)
+			// 2. When it exceeds the paging frame (Example: When there is only 10 records but the second page is called with a page size of 100)
+
+			// link herder at not last page
```
Typo I think
```suggestion
			// link header at not last page
```

JoelSpeed commented on providers/github.go:
```diff
@@ -139,36 +140,95 @@ func (p *GitHubProvider) hasOrgAndTeam(accessToken string) (bool, error) {
+
+			// link herder at not last page
+			// <https://api.github.com/user/teams?page=1&per_page=100>; rel="prev", <https://api.github.com/user/teams?page=1&per_page=100>; rel="last", <https://api.github.com/user/teams?page=1&per_page=100>; rel="first"
+			// link herder at last page (doesn't exist last info)
```
Typo again?
```suggestion
			// link header at last page (doesn't exist last info)
```

JoelSpeed commented on providers/github.go:
```diff
@@ -139,36 +140,95 @@ func (p *GitHubProvider) hasOrgAndTeam(accessToken string) (bool, error) {
+			// link herder at last page (doesn't exist last info)
+			// <https://api.github.com/user/teams?page=3&per_page=10>; rel="prev", <https://api.github.com/user/teams?page=1&per_page=10>; rel="first"
+
+			logger.Printf("1:endpoint :[%s]",endpoint.String())
```
Not sure we need to log these? Could become very noisy in the logs as these would both be printed on every request

JoelSpeed commented on providers/github.go:
```diff
@@ -139,36 +140,95 @@ func (p *GitHubProvider) hasOrgAndTeam(accessToken string) (bool, error) {
+				logger.Printf("3:doesn't exist last page info. ")
+			}else{
+				last = i
+				logger.Printf("3:last page index  :[%d]", last)
```
Potentially over-logging here

toshi-miura replied:
Thank you for the review. How about fixing it?

JoelSpeed commented on providers/github.go:
```diff
@@ -139,36 +140,90 @@ func (p *GitHubProvider) hasOrgAndTeam(accessToken string) (bool, error) {
+			if converr != nil{
+			}else{
+				last = i
+			}
```
I would suggest we  make this just
```
if converr == nil {
  last = i 
}
```
and have no else case, WDYT?

toshi-miura replied:
Thank you for the review.
Corrected as pointed out and added comments

toshi-miura commented:
@JoelSpeed 

>Please rebase on master and add a note to the changelog for the entry and we can then get this merged!

My statement was rebase, except for changes to the original author.

I didn't edit CHANGELOG.md.
Since # 102 was taken over and the change difference of CHANGELOG.md was large

syscll commented on providers/github.go:
```diff
@@ -139,36 +140,89 @@ func (p *GitHubProvider) hasOrgAndTeam(accessToken string) (bool, error) {
+		resp.Body.Close()
+		if err != nil {
+			return false, err
+		}
```
For readability, I would move this code directly after attempting to read `resp.Body` on L174

```
if err != nil {
        resp.Body.Close()
        return false, err
}
resp.Body.Close()
```

JoelSpeed commented:
@toshi-miura I appreciate the diff was large, but you can add a new changelog comment now there's no merge to handle right? Since this is a functional change it needs a changelog entry before it can be merged

toshi-miura commented:
@JoelSpeed 

> it needs a changelog entry 

Does it mean to add to CHANGELOG.md?
Is it like the following?

>Changes since v4.0.0
>#227 Add Keycloak provider (@Ofinka)
>#273 Support Go 1.13 (@dio)
>#274 XXXXXXXXXXXXX (@XXXX)

JoelSpeed commented:
> Is it like the following?

Yep that's exactly what I'd like you to add, thank you :)

toshi-miura commented:
Can't pull the latest master CHANGELOG.md because I am a child of # 102, is that okay?

toshi-miura commented:
I would like to ask you what kind of cycle is it to make a release binary?
In my job, I'm wondering whether to use what I built or use this release build.

toshi-miura commented:
I ’m doing a manual test.
In my environment, there are only about 50 github teams, so I have the following two types of tests

- Test with this source code (Paging logic doesn't work )
- It's bad, but test with editing source code ("per_page": {"5"}, (Paging logic  work ))

syscll commented on CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@
 - [#227](https://github.com/pusher/oauth2_proxy/pull/227) Add Keycloak provider (@Ofinka)
 - [#273](https://github.com/pusher/oauth2_proxy/pull/273) Support Go 1.13 (@dio)
 - [#275](https://github.com/pusher/oauth2_proxy/pull/275) docker: build from debian buster (@syscll)
+- [#274](https://github.com/pusher/oauth2_proxy/pull/274)  Supports many github teams with api pagination support (@toshi-miura ,@apratina)
```
```suggestion
- [#274](https://github.com/pusher/oauth2_proxy/pull/274)  Supports many github teams with api pagination support (@toshi-miura, @apratina)
```

toshi-miura replied:
Thanks for the review, fixed

JoelSpeed commented on CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@
 - [#227](https://github.com/pusher/oauth2_proxy/pull/227) Add Keycloak provider (@Ofinka)
 - [#273](https://github.com/pusher/oauth2_proxy/pull/273) Support Go 1.13 (@dio)
 - [#275](https://github.com/pusher/oauth2_proxy/pull/275) docker: build from debian buster (@syscll)
+- [#274](https://github.com/pusher/oauth2_proxy/pull/274) Supports many github teams with api pagination support (@toshi-miura,@apratina)
```
There should be a space after the comma ideally, sorry for being pedantic
```suggestion
- [#274](https://github.com/pusher/oauth2_proxy/pull/274) Supports many github teams with api pagination support (@toshi-miura, @apratina)
```

toshi-miura commented:
I'm sorry to be late

toshi-miura commented:
> Once the changelog conflict is resolved this should be good to merge.

I resolved the conflict.

JoelSpeed commented:
I'm happy with this now, I've fixed up the changelog again, @syscll gonna override your requested changes as I can see this is now resolved

toshi-miura commented:
@JoelSpeed 

Thank you for the merge.

If the release schedule is decided,
I would be grateful if you would tell the release schedule

syscll commented:
We are planning to do a release in the next few weeks. We're just finalising a few other PRs.

toshi-miura commented:
Thank you very mush!
